spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java

```diff
@@ -977,6 +987,32 @@ private synchronized void clearState(final File fileToWriteTo, final FileState s
 		}
 	}
 
+	private void doFlush(Map<String, FileState> toRemove) {
+		Map<String, FileState> toRestore = new HashMap<>();
+		boolean interrupted = false;
+		for (Entry<String, FileState> entry : toRemove.entrySet()) {
+			if (!interrupted && entry.getValue().close()) {
+				if (FileWritingMessageHandler.this.logger.isDebugEnabled()) {
+					FileWritingMessageHandler.this.logger.debug("Flushed: " + entry.getKey());
+				}
+			}
+			else { // interrupted (stop), re-add
+				interrupted = true;
+				toRestore.put(entry.getKey(), entry.getValue());
```

Comment:
OK. I guess we got here from the `InterruptedException` during `this.flushTask.cancel(true);` in the `stop()`.
Shouldn't we wrap this `put()` to the `synchronized (this) {` as well?
Or `new Flusher().run()` won't be called until the `this.flushTask.cancel(true);` finishes properly and all the puts have happened?

Thanks

Reply:
?? this is just putting to a local variable; we grab the lock below to restore the values to the field.

Reply:
Doh. I see now. Confused with `put` here and `put` there. Never mind.
Although anyway: can't this `this.fileStates.putIfAbsent` be called already after that `Flusher.run()`?

I mean aren't there a situation when we may still have some entries even after `stop()`.

Thanks

Reply:
It's true that if the thread calling stop() is interrupted, there could be some left; I'll give it some more thought.

Reply:
Maybe `Semaphore` to permit `Flusher` to clean up ?
Just a thought...